File: src/main/java/org/dementhium/event/impl/PlayerRestorationTick.java
package org.dementhium.event.impl;

import org.dementhium.event.Tickable;
import org.dementhium.model.player.Player;
import org.dementhium.model.player.Skills;
import org.dementhium.net.ActionSender;
/**
 * 
 * @author 'Mystic Flow
 * 
 */
public class PlayerRestorationTick extends Tickable {

	private Player player;

	private int specialRestoration = 50;
	private int runEnergyRestoration = 2;
	private int levelNormalizationTick = 90;
	private int healTick = 10;

	public PlayerRestorationTick(Player player) {
		super(1);
		this.player = player;
	}

	@Override
	public void execute() {
		if(player.isOnline()) {
			if(player.getSpecialAmount() < 1000) {
				if(specialRestoration > 0) {
					specialRestoration--;
				}
				if(specialRestoration == 0) {
					specialRestoration = 50;
					if(player.getSpecialAmount() + 100 > 1000) {
						player.setSpecialAmount(1000, true);
					} else {
						player.setSpecialAmount(player.getSpecialAmount() + 100, true);
					}
				}
			}
			if(player.getWalkingQueue().getRunEnergy() < 100) {
				if(runEnergyRestoration > 0) {
					runEnergyRestoration--;
				}
				if(runEnergyRestoration == 0) {
					runEnergyRestoration = 2;
					if(!player.getWalkingQueue().isRunningMoving()) {
						player.getWalkingQueue().setRunEnergy(player.getWalkingQueue().getRunEnergy() + 1);
						ActionSender.sendRunEnergy(player);
					}
				}
			}
			if(levelNormalizationTick > 0) {
				levelNormalizationTick--;
			}
			if(levelNormalizationTick == 0) {
				levelNormalizationTick = 90;
				for (int i = 0; i < Skills.SKILL_COUNT; i++) {
					if(i == Skills.HITPOINTS) {
						continue;
					}
					int currentLevel = player.getSkills().getLevel(i);
					int level = player.getSkills().getLevelForXp(i);
					if(currentLevel < level) {
						if(i != Skills.PRAYER && i != Skills.SUMMONING) {
							currentLevel++;
						}
					} else if(currentLevel > level) {
						currentLevel--;
					}
					player.getSkills().setLevel(i, currentLevel);
				}
			}
			if(healTick > 0) {
				healTick--;
			}
			if(healTick == 0) {
				healTick = 10;
				if(player.getSkills().getHitPoints() < player.getSkills().getLevelForXp(Skills.HITPOINTS) * 10) {
					player.getSkills().heal(1);
				}
			}
		} else {
			this.stop();
		}
	}

}

File: src/main/java/org/dementhium/event/impl/PlayerRestorationTick.kt
package org.dementhium.event.impl

import org.dementhium.event.Tickable
import org.dementhium.model.player.Player
import org.dementhium.model.player.skills.Skills
import org.dementhium.model.player.skills.SkillId
import org.dementhium.net.ActionSender

data class PlayerRestorationConfig(
    val specialTimer: Int = 50,
    val runEnergyTimer: Int =  2,
    val levelNormalizationTimer: Int = 90,
    val passiveHealTimer: Int = 10,
    val maxSpecialAmount: Int = 1_000,
    val maxRunEnergy: Int = 100,
    val specialRestoreMax: Int = 100,
    val runEnergyRestoration: Int = 1,
    val levelNormalizationAmount: Int = 1,
    val healAmount: Int = 1,
) {
    companion object {
        val Default = PlayerRestorationConfig()
    }
}

class PlayerRestorationTick(
    private val player: Player,
    private val restorationConfig: PlayerRestorationConfig
) : Tickable(1) {

    constructor(player: Player): this(player, PlayerRestorationConfig.Default)

    private var specialTimer = restorationConfig.specialTimer
    private var runEnergyTimer = restorationConfig.runEnergyTimer
    private var levelNormalizationTimer = restorationConfig.levelNormalizationTimer
    private var passiveHealTimer = restorationConfig.passiveHealTimer

    override fun execute() {
        val isOnline = player.isOnline
        if (!isOnline) return stop()
        if (player.specialAmount < restorationConfig.maxSpecialAmount) {
            if (specialTimer-- == 0) {
                specialTimer = restorationConfig.specialTimer
                player.specialAmount = (player.specialAmount + restorationConfig.specialRestoreMax)
                    .coerceAtMost(restorationConfig.maxSpecialAmount)
            }
        }
        if (player.walkingQueue.runEnergy < restorationConfig.maxRunEnergy) {
            if (runEnergyTimer-- == 0) {
                runEnergyTimer = restorationConfig.runEnergyTimer
                if (!player.walkingQueue.isRunningMoving) {
                    player.walkingQueue.runEnergy += restorationConfig.runEnergyRestoration
                    ActionSender.sendRunEnergy(player)
                }
            }
        }

        if (levelNormalizationTimer-- == 0) {
            levelNormalizationTimer = restorationConfig.levelNormalizationTimer
            for (i in 0 until Skills.SKILL_COUNT) {
                val skillId = SkillId.entries[i]
                val skill = player.skills.getSkill(skillId)
                if (skill.currentLevel == skill.maxLevel) continue
                player.skills.normalizeLevel(skillId, restorationConfig.levelNormalizationAmount) {
                    skill.currentLevel < skill.maxLevel
                }
            }
        }

        if (passiveHealTimer-- == 0) {
            passiveHealTimer = restorationConfig.passiveHealTimer
            if (player.skills.hitPoints < player.skills.maxHitpoints) {
                player.skills.heal(restorationConfig.healAmount)
            }
        }

    }
}